src/Test/Perf/util.csx

```diff
@@ -0,0 +1,180 @@
+        from fileName in allCsiRecursive(childDir, skip)
+        select fileName;
+    return childFiles.Concat(grandChildren).Where((s) => !skip.Contains(s));
+}
+
+/// Runs the script at fileName and returns a task containing the
+/// state of the script.
+public async Task<ScriptState<int>> RunFile(string fileName)
+{
+    var scriptOptions = ScriptOptions.Default.WithFilePath(fileName);
+    var text = File.ReadAllText(fileName);
+    var prelude = "public System.Collections.Generic.List<string> Args = null;";
+    var state = await CSharpScript.RunAsync(prelude);
+    var args = state.GetVariable("Args");
+    args.Value = Args;
+    return await state.ContinueWithAsync<int>(text, scriptOptions);
+}
+
+//
+// Process spawning and error handling.
+//
+
+public class ProcessResult
+{
+    public string ExecutablePath {get; set;}
+    public string Args {get; set;}
+    public int Code {get; set;}
+    public string StdOut {get; set;}
+    public string StdErr {get; set;}
+
+    public bool Failed  => Code != 0;
+    public bool Succeeded  => !Failed;
+}
+
+public ProcessResult ShellOut(
+        string file,
+        string args,
+        CancellationToken? cancelationToken = null)
+{
+    var tcs = new TaskCompletionSource<ProcessResult>();
+    var startInfo = new ProcessStartInfo(file, args);
+    startInfo.RedirectStandardOutput = true;
+    startInfo.RedirectStandardError = true;
+    startInfo.UseShellExecute = false;
+    var process = new Process
+    {
+        StartInfo = startInfo,
+        EnableRaisingEvents = true,
+    };
+
+    if (cancelationToken != null) {
+        cancelationToken.Value.Register(() => process.Kill());
+    }
+
+    process.Exited += (s, a) => {
+        var result = new ProcessResult {
+            ExecutablePath = file,
+            Args = args,
+            Code = process.ExitCode,
+            StdOut = process.StandardOutput.ReadToEnd(),
+            StdErr = process.StandardError.ReadToEnd(),
+        };
+        tcs.SetResult(result);
+        process.Dispose();
+    };
+
+    process.Start();
+
+    return tcs.Task.GetAwaiter().GetResult();
+}
+
+//
+// Timing and Testing
+//
+
+public long WalltimeMs<R>(out R result, Func<R> action)
+{
+    var stopwatch = new Stopwatch();
+    stopwatch.Start();
+    result = action();
+    return stopwatch.ElapsedMilliseconds;
+}
+
+
+//
+// Reporting and logging
+//
+
+private List<Tuple<object, string>> Metrics = new List<Tuple<object, string>>();
+
+/// Logs a message.
+public void Log(string info)
+{
+    System.Console.WriteLine(info);
+}
+
+public void LogProcessResult(ProcessResult result) {
+    Log(String.Format("The process \"{0}\" {1} with code {2}",
+        $"{result.ExecutablePath} {result.Args}",
+        result.Failed ? "failed" : "succeeded",
+        result.Code));
+    Log($"Standard Out:\n{result.StdOut}");
+    Log($"Standard Error:\n{result.StdErr}");
+}
+
+public void Report(object value, string description)
```

Comment:
I'd prefer to see the tests always populate the Metrics and Report could throw if no Metrics had been added (so there's a consistent way of expression the metrics in all the tests, not some that do it at the Report call and some that do it by populating the list).

I imagine there will be a more standard "Metrics" that lists process names and metrics to pull from the .etl traces, and a "ValueMetrics" (or something) as you have here to record things like binary sizes, etc.